### ISSUE TITLE
test/evp_test.c: PKEYKDF / KDF out of memory error not properly detected

### ISSUE THREAD

**Issue by levitte** [openssl/openssl#16519]
These evp_test stanza claim to check for "out of memory":

https://github.com/openssl/openssl/blob/6929c8fb5b46c9c2a383a7c212ee052e0dcef021/test/recipes/30-test_evp_data/evpkdf_scrypt.txt#L56-L63
https://github.com/openssl/openssl/blob/6929c8fb5b46c9c2a383a7c212ee052e0dcef021/test/recipes/30-test_evp_data/evppkey_kdf_scrypt.txt#L56-L63

However, they do so by checking if the result buffer has a mismatch, which of course is plausible, but not how the PKEYKDF test works:

https://github.com/openssl/openssl/blob/6929c8fb5b46c9c2a383a7c212ee052e0dcef021/test/evp_test.c#L2764-L2787

As you can see, if `EVP_PKEY_derive()` fails, which is also plausible for an out of memory situation, the error set is `KDF_DERIVE_ERROR`, not the assumed `KDF_MISMATCH`.

In comes VMS, where every user gets a memory quota:

```
        # INFO: 
        # Reading USER:[LEVITTE.TEST.openssl-3_0_0-beta3-dev.test.recipes.30-test_evp_data]evpkdf_scrypt.txt
        # INFO: 
        # Starting "Scrypt tests (from draft-josefsson-id-scrypt-kdf-03 and others)" tests at line 14
        # INFO: 
        # USER:[LEVITTE.TEST.openssl-3_0_0-beta3-dev.test.recipes.30-test_evp_data]evpkdf_scrypt.txt:57: Expected KDF_MISMATCH got KDF_DERIVE_ERROR
        # C00E6B7B00000000:error:030C0100:digital envelope routines:(unknown function):malloc failure:USER:[LEVITTE.TEST.openssl-3_0_0-beta3-dev.providers.implementations.kdfs]scrypt.c;5:490:
        # INFO: 
        # Completed 6 tests with 1 errors and 0 skipped


pipe MCR dsa10:[ia64.perl-5_20.][000000]perl.exe;1 [--.util]wrap.pl MCR [--.test]evp_test.exe -config "USER:[LEVITTE.TEST.openssl-3_0_0-beta3-dev.test]default-and-legacy.cnf" "USER:[LEVITTE.TEST.openssl-3_0_0-beta3-dev.test.recipes.30-test_evp_data]evpkdf_scrypt.txt" => 2
```

This is a real out of memory error occuring, and we *don't* detect it, even though the stanza comments claims we do, 'cause most machines today have a lot more memory available, 'cause *pfff*, what's 1073743872 bytes of memory these days?

Dunno exactly how to solve this...

**Comment by levitte** [comment]
To be clear, the memory allocation problem occurs here:

https://github.com/openssl/openssl/blob/6929c8fb5b46c9c2a383a7c212ee052e0dcef021/providers/implementations/kdfs/scrypt.c#L488-L492

**Comment by paulidale** [comment]
We could use `setrlimit(2)` to force an out of memory situation on those platforms than have it.
Communicating this back to `evp_test` so the correct error is checked for seems tricky.
How large is the VMS default memory quota?

**Comment by levitte** [comment]
I went looking a bit further just now.  I didn't take note of `maxmem` last evening.

https://github.com/openssl/openssl/blob/09126044f0120b0039b13365b30465d4b6d0f8f0/crypto/evp/pbe_scrypt.c#L154-L155

For 3.0, we apparently changed that:

https://github.com/openssl/openssl/blob/6929c8fb5b46c9c2a383a7c212ee052e0dcef021/providers/implementations/kdfs/scrypt.c#L95-L105

So it seems to me like the stanza needs to change accordingly, probably by increasing *N*

**Comment by levitte** [comment]
Fix in #16521

**Comment by levitte** [comment]
I just tested the changes in #16521 on VMS...  They pass
